From 97444c803d99b45fde78f0d6f814bdebd1378309 Mon Sep 17 00:00:00 2001
From: Ting Li <ting.li@intel.com>
Date: Thu, 23 Jul 2015 18:05:29 +0800
Subject: [PATCH] input: only send a syn event when needed in
 input_dev_release_keys

On input_dev_release_keys, even if there is no key need to release,
it still send a spurious SYN evnet.
In this case, if user poll the event device with EPOLLWAKEUP enable,
there will always be wakeup event to stop system enter suspend

Tracked-On: https://jira01.devtools.intel.com/browse/OAM-105
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-78

Change-Id: I2210491f2e89992debcf40ae7b2b60f2e0060b58
Signed-off-by: Ting Li <ting.li@intel.com>
---
 drivers/input/input.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/input/input.c b/drivers/input/input.c
index 29ca0bb4f561..ebf58cfbf1ae 100644
--- a/drivers/input/input.c
+++ b/drivers/input/input.c
@@ -668,15 +668,18 @@ EXPORT_SYMBOL(input_close_device);
 static void input_dev_release_keys(struct input_dev *dev)
 {
 	int code;
+	bool need_sync = false;
 
 	if (is_event_supported(EV_KEY, dev->evbit, EV_MAX)) {
 		for (code = 0; code <= KEY_MAX; code++) {
 			if (is_event_supported(code, dev->keybit, KEY_MAX) &&
 			    __test_and_clear_bit(code, dev->key)) {
 				input_pass_event(dev, EV_KEY, code, 0);
+				need_sync = true;
 			}
 		}
-		input_pass_event(dev, EV_SYN, SYN_REPORT, 1);
+		if (need_sync)
+			input_pass_event(dev, EV_SYN, SYN_REPORT, 1);
 	}
 }
 
-- 
1.9.1

